code/modules/world_topic/queue_status.dm

```diff
@@ -0,0 +1,36 @@
+/datum/world_topic_handler/queue_status
+	topic_key = "queue_status"
+	requires_commskey = TRUE
+
+// This topic is sent every 10 seconds from the bouncer
+/datum/world_topic_handler/queue_status/execute(list/input, key_valid)
+	var/ckey_check = input["ckey_check"]
+
+	if(!ckey_check)
+		return json_encode(list("error" = "No ckey supplied"))
+
+	var/list/output_data = list()
+	output_data["queue_enabled"] = SSqueue.queue_enabled
+
+	// Decide whether we should hold the player in queue
+	// NOTE: We only queue never seen before players
+	if(SSqueue.queue_enabled)
+		// If they are in the bypass list, let em in
+		if(ckey_check in SSqueue.queue_bypass_list)
+			output_data["allow_player"] = TRUE
+		else // Otherwise
+			// If we have more than the threshold, queue
+			if(length(GLOB.clients) > SSqueue.queue_threshold)
+				output_data["allow_player"] = FALSE
+			else // We have less than the threshold, allow if were in the timeframe
+				if(world.time > (SSqueue.last_letin_time + 3 SECONDS))
+					output_data["allow_player"] = TRUE
+					SSqueue.last_letin_time = world.time
+				else
+					output_data["allow_player"] = FALSE
```

Comment:
Going to point out the one quirk with this logic.

If 2 new players want to connect, it only allows 1 new person in every 3 seconds, regardless of client count, due to how the logic works.

- If people > threshold, disallow
- If people < threshold, allow once every 3 seconds. This is to account for connection latency. If we drop below the threshold by 1, I dont want to instantly funnel 30 people in.

This prevents a mass wave from being let in at once and does spread the load out a bit.

Reply:
Have you checked how long it takes from the time this response is sent to the client actually being added to `GLOB.clients`? If it's potentially way more than those 3 seconds (I don't know how this whole flow works) you could still be overflowing by a bunch of players. 

This is a simple solution that's probably good enough, but a safer way would be to just make a list of clients that have been allowed in but haven't connected yet and add its size to the check on line 23. Then on /mob/new (or something like that) you can just remove the client from the new list. It's tiny overhead and it lets you skip the 3 second wait. The list should probably be filtered for clients that have been connecting for a sensible amount of time, just in case someone is allowed entry but doesn't end up connecting for some reason.

Reply:
> Have you checked how long it takes from the time this response is sent to the client actually being added to GLOB.clients? If it's potentially way more than those 3 seconds (I don't know how this whole flow works) you could still be overflowing by a bunch of players.

Roughly three seconds for a BYOND client to reconnect in my own testing

> This is a simple solution that's probably good enough, but a safer way would be to just make a list of clients that have been allowed in but haven't connected yet and add its size to the check on line 23. Then on /client/new (or something like that) you can just remove the client from the new list. It's tiny overhead and it lets you skip the 3 second wait. The list should probably be filtered for clients that have been connecting for a sensible amount of time, just in case someone is allowed entry but doesn't end up connecting for some reason.

This is already handled by the bouncer. If you pass the queue but fail to connect, you dont have to re-queue, you just get put in, unless you miss your 10 **minute** window, in which case, thats your own fault.

> This solution would also make the queue feasible to run 24/7, because the 3 second wait will be really bad for higher player counts connecting at the same time.

The queue only applies to never seen before players, and bouncing out new joins is something I am ok with.

Reply:
How does that work with map restarts and auto-reconnecting with a lot of players? Will there be enough time for everyone to join?

Reply:
> How does that work with map restarts and auto-reconnecting with a lot of players? Will there be enough time for everyone to join?

- Existing players are not put into the queue
- Map restarts do not have a larger/smaller downtime penality, its the same